feat(ui): add cropping tool (#55)

Spiffyk opened the pull request:
Closes #55 
Closes #155


### TODO:

- [x] ~~Better cursor position handling: make it regions around the crop rectangle's edges, not fractions of the sides. Also probably let the regions be slightly outside of the rectangle as well~~
- [x] ~~Move the crop rectangle by dragging it by its middle~~
    - [x] ~~Do not resize the rectangle while clamping~~
- [x] Fix semi-transparent backdrop rendering when the image is landscape (~~how did that happen?~~ – ah, a very trivial typo, oops)
- [x] Make the cropping tool interact with the built-in undo/redo
- [x] Add the ability to reset the crop
- [x] Update all the docs and labels to reflect the new keybind (conflicted with #183 merged in the meantime)

Spiffyk committed:
feat(ui): add cropping tool (#55)

itsTurnip commented:
Hi and thank you for your contribution! I like the way you implemented it! I guess implementing this feature will finally get swappy one of the preferred tool to use in combination with grim. 

> I'm also thinking maybe the rectangle could actually be split into thirds, like so:

 Personally I would prefer a smaller size of thirds, current are big enough for this task. 

Also I would like to suggest to start painting a new rectangle if the user start to draw it outside of the current one - this was expected behaviour for me, but I got resizing current.

Spiffyk commented:
Thanks for the input, @itsTurnip!

> Personally I would prefer a smaller size of thirds, current are big enough for this task.

Do you mean the current state is fine by you or are you proposing to make them even smaller? Currently, they are not thirds, they are actually quarters (as I, too thought thirds were too big).

> Also I would like to suggest to start painting a new rectangle if the user start to draw it outside of the current one - this was expected behaviour for me, but I got resizing current.

Sounds like a good idea. The current behaviour is that when you hold `Ctrl`, you draw a new rectangle, but perhaps making the whole control mouse-only could be better. I'll try what it feels like to me.

itsTurnip commented:
>  are you proposing to make them even smaller? Currently, they are not thirds, they are actually quarters (as I, too, thought thirds were too big).

Yup, that was what I tried to say. 

Actually, I think it should be a fixed size (just for e.g., 15px). It's not really necessary for this zones to be adjustable after the size of the rectangle. Like for a window managers: a zone to start a windows resize usually have a fixed size.

Spiffyk commented:
(I left the `Ctrl` option in for now as well... probably does not hurt to have it and it's not in any way a considerable amount of extra code to maintain)

itsTurnip commented:
> Anyway, in the meantime I changed the behaviour to be so that when dragging outside of the rectangle, you draw a whole new one. I agree that this is probably more intuitive, I like it.

I tested it. Liked this too :+1: 

Another thing that I would suggest is to change border dragging. Here's how it works right now:

https://github.com/user-attachments/assets/e7b45381-f9da-4dad-93f5-f4c394d2a0d0

You see this: when I try to change the size of the current rectangle it resizes to my cursor position. But I would like to start moving the current position of the border, not to start a new one. 

There's also a visual bug I found. When you start moving the border inside the rectangle the cursor changes it's appearance from cross-hair to the angle. It looks fun when you try to lower the width and higher the height: it starts to blink :smile: 

https://github.com/user-attachments/assets/8c618227-902c-4306-b3dd-78054c4ef09e

Spiffyk commented:
Right, relative movement is probably better. Done it now.

The cursor flicker is fixed now as well.

Thanks for your inputs so far!

Spiffyk commented:
Note: I've put TODOs about what I want to definitely have done before I consider this PR finished. It may not be exhaustive – feel free to suggest more fixes/improvements.

jtheoof commented:
Thank you for this @Spiffyk I'm gonna try to review this soon

Niranjan-GopaL commented:
Hey, awesome extension on the tool !! I think a way to specify that the tool always start in crop mode in the config file would be super helpful too. Really nice implementation there !!

Spiffyk commented:
> @Niranjan-GopaL: I think a way to specify that the tool always start in crop mode in the config file would be super helpful too.

That's already supported, see `paint_mode` in the README :slightly_smiling_face:

jtheoof commented:
Thank you for this @Spiffyk really nice effort was put here. I finally had time to look into it. After testing it, my first feedback is that perhaps we should consider a simplified approach. Not allowing the user to "edit" the crop as you did here, which looks a bit odd as the cursor seems to move away the more I edit the crop (see below)

What I suggest instead is to apply the crop on mouse button release (and updating the paint buffer completely) but plugging it with the undo system, so it's easy to go back to previous state and do a new crop if we messed up.

![Screencast From 2025-08-17 13-29-50](https://github.com/user-attachments/assets/9d236e38-79cb-4a85-a068-8ea33b835d2f)

jtheoof commented:
I've update the CI, it's been a while and github actions were not up to date. Can you please rebase when you get a chance? Some code did not pass the `run-clang-tidy -p build` when I check this MR. The CI will complain after rebase if not fixed.

Spiffyk commented:
Hi, sorry about the delay, @jtheoof!

I just rebased the branch, added a missing dependency declaration (build on NixOS would fail without it), and reformatted.

> After testing it, my first feedback is that perhaps we should consider a simplified approach. Not allowing the user to "edit" the crop as you did here, which looks a bit odd as the cursor seems to move away the more I edit the crop (see below)
>
> What I suggest instead is to apply the crop on mouse button release (and updating the paint buffer completely) but plugging it with the undo system, so it's easy to go back to previous state and do a new crop if we messed up.

Right, that sounds like a very reasonable approach to me! I'll go ahead and look into adapting the behaviour.

Spiffyk commented:
So I've reworked the tool quite a bit, merging it into the other paint types (as it should have been all along).  This way, it integrates with the existing undo-redo mechanism.  The editing logic is now gone.

Spiffyk commented:
Right, I changed the keybind from `c` to `z` because #183 was merged and conflicted with this. Feel free to suggest a different keybind as I'm not sure what would be fitting mnenonically.

Spiffyk committed:
feat(ui): cropping tool rework with undo

Spiffyk commented:
I cleaned up a bit and squashed the history, removing the `gio-unix` change (now in #217).

For now, it's two commits, with the state before and after the rework, but I plan on squashing it up into a single commit once we reach definite agreement on how it's supposed to work.
